### DISCUSSION TITLE
Using a skia_safe::core::surface::Surface in a multi-threaded envirnoment

### DISCUSSION THREAD

**Discussion by FranciscoTorrinha** [rust-skia/rust-skia#939] [General]
Hey guys,
So this might be a dumb question but I have ran into an issue using the skia_safe lib in a multi-threaded environment.

The current state is the following:
I have created an abstraction over skia_safe and I am now trying to use it an external context which implements Send + Sync. However rust does not allow me to do this since `Surface` does not implement Send + Sync, at first I assumed this could be an issue where `RCHandle` is a single thread ref counter but I noticed that other Structs wrapped around RCHandle are Send + Sync.

My questions here are:
- What is the reason behind  `Surface` not being Send + Sync
- If the reason does not break the lib, is there any way I can use the `Surface` as Send + Sync

Thank you for your time,
Francisco

**Comment by FranciscoTorrinha** [comment]
Never mind, the issue was on my part, all I had to do is implement Send + Sync for the struct which was holding the surface 
I'm sorry for the inconvenience

**Reply by pragmatrix** [reply]
Because RCHandle wrappers can be cloned at any time, making and RCHandle Send + Sync would require a mutable wrapped type to implement complete thread safety, meaning that every call must be protected by Mutexes, which Skia does not. 

RCHandle wrapped types _can_ be sent to another thread, but only at times their reference count is one. To make this possible skia-safe, implements the [`ConditionallySend`](https://github.com/rust-skia/rust-skia/blob/d8ff5284700778394554a59f84aceec0ce03d828/skia-safe/src/prelude.rs#L980-L992) trait.

I hope this helps to clarifies why `Surface` and other RCHandle types that wrap mutable Skia types do not implement `Send` nor `Sync`.
